examples/calculator/src/lib.rs

```diff
@@ -0,0 +1,120 @@
+use std::fmt::{self};
+use std::io::{BufRead, Write};
+use std::str::FromStr;
+
+pub struct Calculator {
+  command_buffer: Vec<CalculatorCommand>
+}
+
+pub enum CalculatorOperation { Add, Subtract }
+pub enum CalculatorCommand { Add, Minus, Number(i32) }
+pub enum CalculatorPushResponse { Success, Failure(String) }
+
+impl fmt::Display for CalculatorPushResponse {
+  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+    match self {
+      &CalculatorPushResponse::Success => write!(f, "Success\n"),
+      &CalculatorPushResponse::Failure(ref val) => write!(f, "Failure({})\n", val)
+    }
+  }
+}
+
+
+impl Calculator {
+  pub fn new() -> Calculator {
+    Calculator { command_buffer: Vec::new() }
+  }
+
+  pub fn clear(&mut self) {
+    self.command_buffer.clear()
+  }
+
+  pub fn push_command(&mut self, command: CalculatorCommand) -> CalculatorPushResponse {
+    match command {
+      e @ CalculatorCommand::Number(_) => {
+        if self.command_buffer.len() == 0 {
+          CalculatorPushResponse::Failure("Input operator first".to_owned())
+        } else {
+          self.command_buffer.push(e);
+          CalculatorPushResponse::Success
+        }
+      },
+      e @ _ => {
+        self.command_buffer.push(e);
+        CalculatorPushResponse::Success
+      }
+    }
+  }
+
+  pub fn evaluate(&mut self) -> i32 {
+    let state = (CalculatorOperation::Add, None);
+    let (_, result) = self.command_buffer.iter().fold(state, |(op, val), cmd| {
+      match cmd {
+        &CalculatorCommand::Number(num) => {
+          let new_val = val.map(|v| match op {
+            CalculatorOperation::Add => v + num,
+            CalculatorOperation::Subtract => v - num
+          }).or(Some(num));
+          (op, new_val)
+        },
+        &CalculatorCommand::Add => (CalculatorOperation::Add, val),
+        &CalculatorCommand::Minus => (CalculatorOperation::Subtract, val)
+      }
+    });
+
+    self.clear();
+    result.unwrap_or(0)
+  }
+}
+
+pub struct RWCalculator<R: BufRead, W: Write> {
```

Comment:
This RW calculator is a whole lotta fluff in order to make the binary for this crate actually do something. Its not tested, although it could be via worlds & tags.